File: be-scraper-fastapi/app/models/competition.py
from sqlmodel import SQLModel, Field
from sqlalchemy import Column
from sqlalchemy.types import JSON
from typing import Optional
from datetime import datetime, timezone
import uuid

class Competition(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    deleted_at: Optional[datetime] = None
    image_path: Optional[str] = None
    # is_open: bool
    tk_number: Optional[str] = None # used in the Teams page
    t3kys_number: Optional[str] = None # used in the t3kys platform
    application_link_tr: Optional[str] = None
    application_link_en: Optional[str] = None
    application_link_ar: Optional[str] = None
    tr_name: Optional[str] = None
    tr_description: Optional[str] = None
    tr_link: Optional[str] = None
    en_name: Optional[str] = None
    en_description: Optional[str] = None
    en_link: Optional[str] = None
    ar_name: Optional[str] = None
    ar_description: Optional[str] = None
    ar_link: Optional[str] = None
    years: list[str] = Field(default_factory=list, sa_column=Column(JSON)) # The years that the competition is held
    min_member: Optional[int] = None # Minimum number of members in a team
    max_member: Optional[int] = None # Maximum number of members in a team
    ggroup: Optional[str] = None # competition google group link
    # comments: list[uuid.UUID]

class CompetitionData(SQLModel, table=True):
    competition_id: int = Field(primary_key=True)
    year: str = Field(primary_key=True)
    timeline: Optional[dict] = Field(default_factory=dict, sa_column=Column(JSON)) # timeline dates in JSON format (takvim)
    awards: Optional[dict] = Field(default_factory=dict, sa_column=Column(JSON)) # awards details in JSON format
    criteria: Optional[dict] = Field(default_factory=dict, sa_column=Column(JSON)) # evaluation criteria files links in JSON format

class Report_File(SQLModel, table=True):
    id: Optional[uuid.UUID] = Field(default_factory=uuid.uuid4, primary_key=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    deleted_at: Optional[datetime] = None
    competition_id: int
    team_id: Optional[uuid.UUID] = None
    year: str
    file_path: str
    rank: Optional[str] = None  # "finalist", "derece"
    stage : Optional[str] = None  # "critical-design", "pre-assessment", "final-assessment", "final-presentation"
    language: Optional[str] = None  # "tr", "en", "ar"

class Result_File(SQLModel, table=True):
    id: Optional[uuid.UUID] = Field(default_factory=uuid.uuid4, primary_key=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    deleted_at: Optional[datetime] = None
    competition_id: int
    year: str
    stage: str # "final", "pre-assessment"
    file_path: str

File: be-main-hono/src/types/competition.ts
export interface Competition {
	id: number;
	created_at: string;
	updated_at: string;
	deleted_at: string | null;
	image_path: string | null;
	tk_number: string | null;
	t3kys_number: string | null;
	application_link_tr: string | null;
	application_link_en: string | null;
	application_link_ar: string | null;
	tr_name: string | null;
	tr_description: string | null;
	tr_link: string | null;
	en_name: string | null;
	en_description: string | null;
	en_link: string | null;
	ar_name: string | null;
	ar_description: string | null;
	ar_link: string | null;
	years: number[];
	min_member: number | null;
	max_member: number | null;
	ggroup: string | null;
}

export interface TimelineEntry {
	description: string;
	date: string;
}

export interface AwardEntry {
	rank: string;
	prize: string;
}

export interface CompetitionData {
	competition_id: number;
	year: number;
	timeline: TimelineEntry[] | null;
	awards: Record<string, AwardEntry[]> | AwardEntry[] | null;
	criteria: Record<string, any> | null;
}

export interface CreateCompetitionPayload {
	image_path?: string | null;
	tk_number?: string | null;
	t3kys_number?: string | null;
	application_link_tr?: string | null;
	application_link_en?: string | null;
	application_link_ar?: string | null;
	tr_name?: string | null;
	tr_description?: string | null;
	tr_link?: string | null;
	en_name?: string | null;
	en_description?: string | null;
	en_link?: string | null;
	ar_name?: string | null;
	ar_description?: string | null;
	ar_link?: string | null;
	years?: number[];
	min_member?: number | null;
	max_member?: number | null;
	ggroup?: string | null;
}

export type UpdateCompetitionPayload = Partial<CreateCompetitionPayload>;

export const mapCompetitionRow = (row: any): Competition => {
	return {
		id: Number(row.id),
		created_at: row.created_at,
		updated_at: row.updated_at,
		deleted_at: row.deleted_at ?? null,
		image_path: row.image_path ?? null,
		tk_number: row.tk_number ?? null,
		t3kys_number: row.t3kys_number ?? null,
		application_link_tr: row.application_link_tr ?? null,
		application_link_en: row.application_link_en ?? null,
		application_link_ar: row.application_link_ar ?? null,
		tr_name: row.tr_name ?? null,
		tr_description: row.tr_description ?? null,
		tr_link: row.tr_link ?? null,
		en_name: row.en_name ?? null,
		en_description: row.en_description ?? null,
		en_link: row.en_link ?? null,
		ar_name: row.ar_name ?? null,
		ar_description: row.ar_description ?? null,
		ar_link: row.ar_link ?? null,
		years: (() => {
			try {
				return row.years ? JSON.parse(row.years) : [];
			} catch {
				return [];
			}
		})(),
		min_member: row.min_member != null ? Number(row.min_member) : null,
		max_member: row.max_member != null ? Number(row.max_member) : null,
		ggroup: row.ggroup ?? null,
	};
};

